pinot/generative/torch_gvae/utils.py

```diff
@@ -0,0 +1,327 @@
+          A* = A + I # adding self loops
+          D* = sum(A*, axis=1) # row sum
+
+    This matrix is inspired by first order approximation of localized spectral filters
+    on graphs.
+
+    To see more details, see Kipf and Welling (2016)
+    Note that I don't have a good name for this matrix and not sure if people have
+    come up with an "official" name for this matrix.
+
+    But  I - D^{-1/2}AD^{-1/2} is the normalized Laplacian matrix
+    And I + D^{-1/2}AD^{-1/2} = D*^{-1/2}A*D*^{-1/2}
+
+    So one can think of D*^{-1/2}A*D*^{-1/2} as the shifted negative version
+    of the normalized Laplacian
+
+    Arg:
+        adj: adjacency matrix (may not be in sparse layout)
+
+    Returns:
+
+
+    """
+    adj = sp.coo_matrix(adj)
+    adj_ = adj + sp.eye(adj.shape[0])
+    D_tilde = np.array(adj_.sum(1))
+    degree_mat_inv_sqrt = sp.diags(np.power(D_tilde, -0.5).flatten())
+    neg_shifted_lap = adj_.dot(degree_mat_inv_sqrt).transpose().dot(degree_mat_inv_sqrt).tocoo()
+    return sparse_mx_to_torch_sparse_tensor(neg_shifted_lap)
+
+
+def sparse_mx_to_torch_sparse_tensor(sparse_mx):
+    """Convert a scipy sparse matrix to a torch sparse tensor."""
+    sparse_mx = sparse_mx.tocoo().astype(np.float32)
+    indices = torch.from_numpy(
+        np.vstack((sparse_mx.row, sparse_mx.col)).astype(np.int64))
+    values = torch.from_numpy(sparse_mx.data)
+    shape = torch.Size(sparse_mx.shape)
+    return torch.sparse.FloatTensor(indices, values, shape)
+
+
+def get_train_test_val_set(raw_data, train_ratio=0.8, test_ratio=0.1):
+    assert (train_ratio + test_ratio < 1)
+    val_ratio = 1 - train_ratio - test_ratio
+    # Returns training_data, validation_data, test_data
+    
+    data = []
+
+    for g, measurement in raw_data:
+        try:
+            adj_mat_sp = g.adjacency_matrix(True)
+
+            adj_mat = adj_mat_sp.to_dense().detach()
+            pos_weight = torch.FloatTensor(
+                    [(adj_mat.shape[0] * adj_mat.shape[0] - adj_mat.sum()) \
+                    / adj_mat.sum()])
+        
+            norm = adj_mat.shape[0] * adj_mat.shape[0] \
+                    / float((adj_mat.shape[0] * adj_mat.shape[0] \
+                        - adj_mat.sum()) * 2)
+
+            data.append((g, adj_mat, pos_weight, adj_mat, measurement))
+
+        except Exception as e:
+            print(e)
+            continue
+
+    N_molecules = len(data)
+
+    N_train = int(N_molecules * train_ratio)
+    N_test = int(N_molecules * test_ratio)
+    N_val  = N_molecules - N_train - N_test
+
+    rand_indices = np.random.permutation(N_molecules)
+    train_indices = rand_indices[:N_train]
+    test_indices  = rand_indices[N_train:N_train+N_test]
+    val_indices   = rand_indices[N_train+N_test:]
+
+    train_data = [data[i] for i in train_indices]
+    test_data  = [data[i] for i in test_indices]
+    val_data   = [data[i] for i in val_indices]
+    return train_data, test_data, val_data
+
+
+def prepare_train_test_val(data):
```

Comment:
could you comment a bit on the necessity of this as opposed to partition / sample small graphs and then batch them together?

Reply:
I don't know the answer to this question yet. But updated code allows us to choose how many molecules get "batched" into a macromolecule per iteration